tests/test_logging_api.py

```diff
@@ -0,0 +1,819 @@
+            assert False, message
+        # Then I see a ERROR message from the std library logger with the
+        # traceback attached
+
+        for record in logger.records:
+            if all((
+                record.levelno == logging.ERROR,
+                record.exc_info,
+                isinstance(record.exc_info[1], AssertionError),
+                record.exc_info[1].args[0] == message
+            )):
+                return
+
+        assert not logger.records, "No matching record found"
+
+    def test_record_runtime_error(self):
+        message = "Woops!"
+        exc_type = RuntimeError
+
+        try:
+            # Given a logger context
+            with TestRomaineLogger() as logger:
+                # When I raise an assertion
+                raise exc_type(message)
+        except exc_type:
+            pass
+
+        # Then I see a ERROR message from the std library logger with the
+        # traceback attached
+        for record in logger.records:
+            if all((
+                    record.levelno == logging.ERROR,
+                    record.exc_info,
+                    isinstance(record.exc_info[1], exc_type),
+                    record.exc_info[1].args[0] == message
+            )):
+                return
+
+        assert False, "No matching record found"
+
+
+class TestLoggingAPIStatistics(unittest.TestCase):
+
+    def _fail(self):
+        assert False
+
+    def _skip(self):
+        raise exc.SkipTest
+
+    def _run_for_steps(self, fn=None):
+        """
+        yields the logger context, then runs fn for each step
+        """
+        # Given a logger context
+        with TestRomaineLogger() as logger:
+            yield logger
+            # And a test feature
+            feature = given_a_feature()
+            # When I enter the feature context
+            with logger.in_feature(feature):
+                # Given each scenario outline in the feature
+                for scenario in feature["elements"]:
+                    # When I enter the scenario outline context
+                    with logger.in_scenario_outline(scenario):
+                        # Given each example in the scenario outline
+                        for example in scenario["examples"]:
+                            # When I enter the example context
+                            with logger.in_scenario_outline_example(example):
+                                # Given each row in the example
+                                for index in range(len(example["hashes"])):
+                                    with logger\
+                                        .in_scenario_outline_example_row(
+                                            example, index) as run:
+                                        for step in run:
+                                            with logger.in_step(step):
+                                                if fn is not None:
+                                                    fn()
```

Comment:
I'd say it would be an issue to do this in production code - for test code I've taken a few shortcuts, and hopefully this should be fine